include/seqan3/range/view/pairwise_combine.hpp

```diff
@@ -0,0 +1,595 @@
+ * \{
+ */
+
+/*!\brief             A view adapter that generates all pairwise combinations of the elements of the source range.
+ * \tparam urng_t     The type of the range being processed. See below for requirements.
+ * \param[in] urange  The range being processed.
+ * \returns           A range over all pairwise combinations. See below for the properties of the returned range.
+ * \ingroup view
+ *
+ * \details
+ *
+ * This view generates two-element tuples representing all possible combinations of the elements of the source range
+ * while ignoring the order of the elements. If the source range has less than two elements the returned range is empty,
+ * otherwise the size of the returned view corresponds to the binomial coefficient `n choose 2`, where `n` is the
+ * size of the source range. The reference type of this range is a tuple over the reference type of the source range.
+ * Constness will not be propagated to the reference types of the source range.
+ *
+ * ### View properties
+ *
+ * | range concepts and reference_t  | `urng_t` (underlying range type)      | `rrng_t` (returned range type)                                       |
+ * |---------------------------------|:-------------------------------------:|:--------------------------------------------------------------------:|
+ * | std::ranges::InputRange         |                                       | *undefined*                                                          |
+ * | std::ranges::ForwardRange       | *required*                            | *preserved*                                                          |
+ * | std::ranges::BidirectionalRange |                                       | *preserved*                                                          |
+ * | std::ranges::RandomAccessRange  |                                       | *preserved*                                                          |
+ * | std::ranges::ContiguousRange    |                                       | *lost*                                                               |
+ * |                                 |                                       |                                                                      |
+ * | std::ranges::ViewableRange      | *required*                            | *guaranteed*                                                         |
+ * | std::ranges::View               |                                       | *guaranteed*                                                         |
+ * | std::ranges::SizedRange         |                                       | *guaranteed* iff urng_t models std::RandomAccessRange                |
+ * | std::ranges::CommonRange        |                                       | *guaranteed*                                                         |
+ * | std::ranges::OutputRange        |                                       | *lost*                                                               |
```

Comment:
Actually I think you are preserving this, because you are reference type can be promoted to an `xvalue`:

```cpp
#include <tuple>

int i = 1;
int j = 2;

std::tuple<int&, int&> foo()
{
    return {i, j};
}

int main()
{
    int k = 3;
    int l = 4;
    foo() = std::tie(k, l);
    return i; // returns 3
}
```

Reply:
This case is explicitly prevented by the WritableConcept! See http://eel.is/c++draft/iterator.concepts#iterator.concept.writable-6

Reply:
:+1:

Although there will be some hack before C++20 so exactly this will work for view::zip. The hack might be a custom type or actually adding const-qualified assignment to tuples of references 😮